#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# PS4 Controller Channel Mapping
#
#     Axis description
# 0   Left X (left/right)
# 1   Left Y (up/down)
# 2   L2
# 3   Right X
# 4   Right Y
# 5   R2
# 6   D Pad X
# 7   D Pad Y
#
#     Button description
# 0   ✕
# 1   ○
# 2   △
# 3   □
# 4   L1
# 5   R1
# 6   L2
# 7   R2
# 8   SHARE
# 9   OPTIONS
# 10  MODE
# 11  Left Stick Click
# 12  Right Stick Click


PACKAGE = "iturov_control"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# PS4 Controller Axis List
axis_enum = gen.enum([
    gen.const("Axis_Disabled",      int_t, -1, ""),
    gen.const("Left_Stick_X",       int_t, 0,  "left/right"),
    gen.const("Left_Stick_Y",       int_t, 1,  "up/down"),
    gen.const("Left_Trigger",       int_t, 2,  ""),
    gen.const("Right_Stick_X",      int_t, 3,  "left/right"),
    gen.const("Right_Stick_Y",      int_t, 4,  "up/down"),
    gen.const("Right_Trigger",      int_t, 5,  ""),
    gen.const("D_Pad_X",            int_t, 6,  "left/right"),
    gen.const("D_Pad_Y",            int_t, 7,  "up/down"),
    gen.const("Trigger_Pair",       int_t, 8,  "right trigger is positive"),
], "An enum to set axis index")

# PS4 Controller Button List
button_enum = gen.enum([
    gen.const("Button_Disabled",    int_t, -1, ""),
    gen.const("X_Button",           int_t, 0,  ""),
    gen.const("Circle_Button",      int_t, 1,  ""),
    gen.const("Triangle_Button",    int_t, 2,  ""),
    gen.const("Square_Button",      int_t, 3,  ""),
    gen.const("L1_Button",          int_t, 4,  ""),
    gen.const("R1_Button",          int_t, 5,  ""),
    gen.const("L2_Button",          int_t, 6,  ""),
    gen.const("R2_Button",          int_t, 7,  ""),
    gen.const("SHARE_Button",       int_t, 8,  ""),
    gen.const("OPTIONS_Button",     int_t, 9,  ""),
    gen.const("MODE_Button",        int_t, 10, ""),
    gen.const("Left_Stick_Click",   int_t, 11, ""),
    gen.const("Right_Stick_Click",  int_t, 12, ""),
], "An enum to set button index")

# Set Axis Parameters
gen.add("X_Axis",           int_t, 0, "", 1,  0, 7,  edit_method=axis_enum)
gen.add("Y_Axis",           int_t, 0, "", 0,  0, 7,  edit_method=axis_enum)
gen.add("Z_Axis",           int_t, 0, "", 8,  0, 7,  edit_method=axis_enum)
gen.add("Roll_Axis",        int_t, 0, "", -1, 0, 7,  edit_method=axis_enum)
gen.add("Pitch_Axis",       int_t, 0, "", 4,  0, 7,  edit_method=axis_enum)
gen.add("Yaw_Axis",         int_t, 0, "", 3,  0, 7,  edit_method=axis_enum)

# Set Button Parameters
gen.add("Arm_Button",       int_t, 0, "", 5,  0, 14, edit_method=button_enum)
gen.add("Disarm_Button",    int_t, 0, "", 4,  0, 14, edit_method=button_enum)
gen.add("Manual_Button",    int_t, 0, "", 0,  0, 14, edit_method=button_enum)
gen.add("Stabilize_Button", int_t, 0, "", 2,  0, 14, edit_method=button_enum)
gen.add("Fast_Mode_Button", int_t, 0, "", 1,  0, 14, edit_method=button_enum)
gen.add("Slow_Mode_Button", int_t, 0, "", 3,  0, 14, edit_method=button_enum)
gen.add("Cam_Tilt_X",       int_t, 0, "", 6,  0, 14, edit_method=axis_enum)
gen.add("Cam_Tilt_Y",       int_t, 0, "", 7,  0, 14, edit_method=axis_enum)
gen.add("Cam_Tilt_Reset",   int_t, 0, "", 11, 0, 14, edit_method=button_enum)

# Set Scaling Parameters
linear_max = 1
angular_max = 1
gen.add("Pub_Rate",          int_t, 0, "the maximum publish rate", 100, 1, 200)
gen.add("Expo",           double_t, 0, "improves control at low levels", 1, 0, 1)
gen.add("X_Scaling",      double_t, 0,  "linear scaling",  1, -linear_max, linear_max)
gen.add("Y_Scaling",      double_t, 0,  "linear scaling",  1, -linear_max, linear_max)
gen.add("Z_Scaling",      double_t, 0,  "linear scaling",  1, -linear_max, linear_max)
gen.add("Roll_Scaling",   double_t, 0, "angular scaling",  1, -angular_max, angular_max)
gen.add("Pitch_Scaling",  double_t, 0, "angular scaling",  1, -angular_max, angular_max)
gen.add("Yaw_Scaling",    double_t, 0, "angular scaling", -1, -angular_max, angular_max)

# Tilting Camera Parameters
gen.add("Cam_Tilt_Speed", int_t, 0, "step size", 50, 0, 100)

exit(gen.generate(PACKAGE, "Joystick Params", "Joy"))
